rp2040-hal/src/spi.rs

```diff
@@ -1,3 +1,184 @@
 //! Serial Peripheral Interface (SPI)
 // See [Chapter 4 Section 4](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
 // TODO
+
+use pac::{RESETS, SPI0, SPI1};
+
+use crate::resets::SubsystemReset;
+
+///
+pub trait Spi {
+    /// Initialize the spi
+    fn init(&mut self, resets: &mut RESETS, baudrate: u32);
+
+    ///
+    fn set_baudrate(&mut self, baudrate: u32) -> u32;
+
+    ///
+    fn set_format(&mut self, data_bits: u8, cpol: Cpol, cpha: Cpha);
+
+    ///
+    fn is_writable(&self) -> bool;
+
+    ///
+    fn is_readable(&self) -> bool;
+
+    ///
+    fn is_busy(&self) -> bool;
+}
+
+///
+pub enum Cpol {
+    /// CPOL=0
+    _0 = 0,
+    /// CPOL=1
+    _1,
+}
+
+///
+pub enum Cpha {
+    ///CPHA=0
+    _0 = 0,
+    ///CPHA=1
+    _1,
+}
+
+macro_rules! impl_spi {
+    ($name:ident, $dev:ident, $description:literal) => {
+        /// Spi 0
+        #[doc()]
+        pub struct $name {
+            spi_dev: $dev,
+        }
+
+        impl $name {
+            /// Create new spi device
+            pub fn new(spi_dev: $dev) -> Self {
+                $name { spi_dev }
+            }
+
+            /// Free underlying spi registers
+            pub fn free(self) -> $dev {
+                self.spi_dev
+            }
+        }
+
+        impl Spi for $name {
+            fn init(&mut self, resets: &mut RESETS, baudrate: u32) {
+                self.spi_dev.reset_bring_down(resets);
+                self.spi_dev.reset_bring_up(resets);
+
+                self.set_baudrate(baudrate);
+                self.spi_dev
```

Comment:
This and the following write to `sspcsr0` clobber the result of `set_baudrate`

Reply:
Ah yes some debugging code.. Should be fixed now